bibliography:biblio.bib

* Data2016_2017
  There are data I used during my last year of PhD.
** git-annex

   In new git repo (git bare repo too)

   #+BEGIN_SRC bash
   git annex init
   #+END_SRC

   Data are on TIMC-BCM-15 and krakenator.

*** How to...
    For more details see: https://git-annex.branchable.com/walkthrough/
**** add krakenator remote
#+BEGIN_SRC bash
git remote add krakenator krakenator.imag.fr:/home/cayek/Project/Data2016_2017
#+END_SRC

**** add data
 #+BEGIN_SRC bash
 git annex add file_name
 #+END_SRC

**** copy data to krakenator
 #+BEGIN_SRC bash
 git annex copy --to krakenator
 #+END_SRC

**** get data from krakenator_rsync
 #+BEGIN_SRC bash
 git annex sync krakenator
 git annex get file_name
 #+END_SRC


* Dataset
  Some informations about dataset in this repo
** ./SSMPG2015/ 
   Dataset simulated by Katie Lotterhos for the school SSMPG2015
** ./MathieuGautier/
   Dataset used in *Genome scan methods against more complex models: when and how much should we trust them?* of piere de villemereuil et al.
*** TODO Convert 
    .csv into .RData
*** =mec12705-sup-0002-Pythonscripts/=
    Python script that generated dataset 
*** Monogenic
    I find this in an older .Rmd in my first LFMM project of 2016
#+BEGIN_SRC R
outlier = c(546) # for monogenic
#+END_SRC

*** Polygenic
    I find this in an older .Rmd in my first LFMM project of 2016
#+BEGIN_SRC R 
outlier = c(2793,1850,583,4083,3349,860,4785,706,947,939,1819,925,403,2867,2897,97,3102,2618,708,1190,2471,1533,3924,2395,2690,2926,1511,668,4826,4755,638,4148,1777,1869,2252,4326,397,3416,3171,2451,1233,2055,3013,3202,1055,3484,2984,2145,4547,4831) + 1
#+END_SRC
** ./AthalianaGegMapLines/
  - Data download from: http://bergelson.uchicago.edu/?page_id=790
  - [[http://bergelson.uchicago.edu/wp-content/uploads/2015/04/call_method_75.tar.gz][download the data]]
  There are data I used in TESS3 second article
  
** ./1000Genomes/
*** ./1000Genomes/Phase3Chrm22/
    Phase 3 version of the 1000 genome, only the chromosom 22. I ddl the vcf
    file [[ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/][here]]
    ddl file: 
#+BEGIN_SRC bash
curl -O ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr22.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
curl -O ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/integrated_call_samples_v3.20130502.ALL.panel
#+END_SRC
** ./GSE42861/
   Données trouvé ici:
   https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE42861
   Utilisé dans ce papier cite:Rahmani_2016.
*** file:./GSE42861/script_methylome.R
    C'est le fichier qu'olivier a utilisé pour pour ddl les données. Ca marche
    pas chez moi...
*** ./GSE42861/exp861.RData
    C'est les variables defs dans file:./GSE42861/script_methylome.R. Output par
    olivier.
*** DONE ./GSE42861/DataPreprocessing.R
    CLOSED: [2017-02-22 mer. 10:54]
    Goal: repruduce data processing of paper cite:Zou_2014 
    
    En faite je vais pas me prendre la tête :D
** ./refractorDemo/
   data found here: https://github.com/cozygene/refactor. There are data used in
   the demo of refractor method.
** ./Hgdp_Li/
   Hgdp data used in cite:frichot13_testin_assoc_between_loci_envir anylisis. I
   I found this dataset on patator.imag.fr.
